### PR TITLE
feat(sumo): Enable the extraction of train-relevant data from SUMO

### PR DESCRIPTION
**Pull request by schwepmo**
## Type of this PR 

- [ ] Bug fix
- [x] Enhancement

## Description

* this PR includes multiple changes related to simulating SUMO trains with MOSAIC
* Firstly, the `additionalData`-field in `VehicleData` has been utilized to be filled with a `TrainData`-object generated by the `SimulationFacade`
* The `TrainData`-class currently contains upcoming stops and a line parameter, which should enable some form of boarding advice
* additionally, a workaround for SUMO scenarios with railway signals has been implemented
* Essentially, railway signals are treated as traffic lights in SUMO, but can be defined without phase-definitions. In this case MOSAIC fails to handle the railway signal. To intercept this an additional null-check was implemented
## Issue(s) related to this PR

* relates to internal issue 510, 528, 536
* 
## Affected parts of the online documentation

* documentation has to be evaluated separately

## Definition of Done

- [x] You have read CONTRIBUTING.md carefully.
- [x] You have signed the [Contributor License Agreement](http://www.eclipse.org/legal/CLA.php).
- [x] Your GitHub user id is linked with your Eclipse Account.
- [x] `origin/main` has been merged into your Fork.
- [x] New functionality is covered by unit tests or integration tests. Code coverage must not decrease.
- [x] There are no new SpotBugs warnings. 
- [x] Coding guidelines have been followed (see CONTRIBUTING.md).
- [x] All tests pass.

### CODE REVIEW

File: fed/mosaic-sumo/src/main/java/org/eclipse/mosaic/fed/sumo/bridge/traci/reader/TypeBasedTraciReader.java [hunk 1]
```diff
@@ -71,6 +74,7 @@ protected Object readFromStream(DataInputStream in) throws IOException {
                 return readUnsignedByte(in);
             case TraciDatatypes.COMPOUND:
                 readInt(in); // required at least for reading inductionloop vehicle data compound
```
Comment (kschrab) [inline fed/mosaic-sumo/src/main/java/org/eclipse/mosaic/fed/sumo/bridge/traci/reader/TypeBasedTraciReader.java:76]: the comment may not fit anymore

File: fed/mosaic-sumo/src/main/java/org/eclipse/mosaic/fed/sumo/bridge/traci/reader/TypeBasedTraciReader.java [hunk 2]
```diff
@@ -38,8 +41,8 @@ public TypeBasedTraciReader(Matcher<Object> matcher) {
         super(matcher);
     }
 
-    public void registerCompoundReader(AbstractTraciResultReader<?> compoundReader) {
-        this.compoundReader = compoundReader;
+    public void registerCompoundReader(Integer command, AbstractTraciResultReader<?> compoundReader) {
```
Comment (kschrab) [inline fed/mosaic-sumo/src/main/java/org/eclipse/mosaic/fed/sumo/bridge/traci/reader/TypeBasedTraciReader.java:44]: use primitive `int` as parameter type

File: fed/mosaic-sumo/src/main/java/org/eclipse/mosaic/fed/sumo/ambassador/AbstractSumoAmbassador.java [hunk 3]
```diff
@@ -1368,7 +1368,11 @@ private void initializeTrafficLights(long time) throws InternalFederateException
         Map<String, Collection<String>> trafficLightGroupLaneMap = new HashMap<>();
         for (String trafficLightGroupId : trafficLightGroupIds) {
             try {
-                trafficLightGroups.add(bridge.getTrafficLightControl().getTrafficLightGroup(trafficLightGroupId));
+                TrafficLightGroup trafficLightGroup = bridge.getTrafficLightControl().getTrafficLightGroup(trafficLightGroupId);
+                if (trafficLightGroup == null) {  // FIXME: workaround for railway signals
```
Comment (kschrab) [inline fed/mosaic-sumo/src/main/java/org/eclipse/mosaic/fed/sumo/ambassador/AbstractSumoAmbassador.java:1372]: remove "FIXME" from comment. Better write "reading traffic light groups does not work for railway signals"

**Review by kschrab — CHANGES_REQUESTED** [review]
Some minor changes requested

File: fed/mosaic-sumo/src/main/java/org/eclipse/mosaic/fed/sumo/bridge/facades/TrafficLightFacade.java [hunk 4]
```diff
@@ -184,7 +184,9 @@ public TrafficLightGroup getTrafficLightGroup(String trafficLightGroupId) throws
                             controlledLinks,
                             junctionPosition
                     );
-
+            if (trafficLights.size() == 0) { // FIXME: workaround for railway signals
```
[the remark this answers is not in the document]
Comment (schwepmo) [reply fed/mosaic-sumo/src/main/java/org/eclipse/mosaic/fed/sumo/bridge/facades/TrafficLightFacade.java:187]: Just to clarify, from SUMO-side this behavior is intended. Railway signals without a logic will simply ensure that there is no more than one train on a track simultaneously.
This means that a possible fix is on our side, as we would have to differentiate between "regular" traffic lights and railway signals when requesting the data from TraCI.

File: fed/mosaic-sumo/src/main/java/org/eclipse/mosaic/fed/sumo/bridge/traci/reader/AbstractSubscriptionTraciReader.java [hunk 5]
```diff
@@ -48,7 +48,8 @@ protected T readFromStream(DataInputStream in) throws IOException {
         for (int i = 0; i < varCount; i++) {
             int varId = readUnsignedByte(in);
             int varStatus = readUnsignedByte(in);
-
+            // TODO: hand over varId
```
[the remark this answers is not in the document]
Comment (schwepmo) [reply fed/mosaic-sumo/src/main/java/org/eclipse/mosaic/fed/sumo/bridge/traci/reader/AbstractSubscriptionTraciReader.java:51]: No, unless we're not satisfied with the current solution.

File: fed/mosaic-sumo/src/main/java/org/eclipse/mosaic/fed/sumo/bridge/traci/reader/TypeBasedTraciReader.java [hunk 6]
```diff
@@ -80,4 +84,8 @@ protected Object readFromStream(DataInputStream in) throws IOException {
                 throw new RuntimeException("Subscribed variable type " + varReturnType + " not known.");
         }
     }
+
+    public void setVar(int varId) {
```
[the remark this answers is not in the document]
Comment (schwepmo) [reply fed/mosaic-sumo/src/main/java/org/eclipse/mosaic/fed/sumo/bridge/traci/reader/TypeBasedTraciReader.java:88]: done.

File: fed/mosaic-sumo/src/main/java/org/eclipse/mosaic/fed/sumo/bridge/traci/reader/TypeBasedTraciReader.java (hunk 1 above)
Comment (schwepmo) [reply fed/mosaic-sumo/src/main/java/org/eclipse/mosaic/fed/sumo/bridge/traci/reader/TypeBasedTraciReader.java:76]: changed to "this field needs to be read but can be ignored"

File: fed/mosaic-sumo/src/main/java/org/eclipse/mosaic/fed/sumo/bridge/facades/TrafficLightFacade.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (kschrab) [reply fed/mosaic-sumo/src/main/java/org/eclipse/mosaic/fed/sumo/bridge/facades/TrafficLightFacade.java]: Is this documented on sumo-side? Would be great if you could link to sumo-doc or describe train-specific behavior in our internal wiki

[the remark this answers is not in the document]
Comment (schwepmo) [reply fed/mosaic-sumo/src/main/java/org/eclipse/mosaic/fed/sumo/bridge/facades/TrafficLightFacade.java]: I added some documentation internally.

**Pull request merged**